tests/0147-consumer_group_consumer_mock.c

```diff
@@ -0,0 +1,682 @@
+        } else if (variation == 1) {
+                /* First HB assigns again
+                 * Second HB acks assignment */
+                expected_heartbeats += 2;
+        }
+
+        TEST_SAY("Awaiting acknowledge heartbeat\n");
+        TEST_ASSERT((found_heartbeats =
+                         wait_all_heartbeats_done(mcluster, expected_heartbeats,
+                                                  100)) == expected_heartbeats,
+                    "Expected %d heartbeats, got %d", expected_heartbeats,
+                    found_heartbeats);
+
+        expected_rebalance_cnt++;
+        rebalance_exp_event = RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS;
+
+        rd_kafka_mock_clear_requests(mcluster);
+        /* Close c, no error is returned */
+        TEST_CALL_ERR__(rd_kafka_consumer_close(c));
+
+        TEST_ASSERT(rebalance_cnt == expected_rebalance_cnt,
+                    "Expected %d rebalance events, got %d",
+                    expected_rebalance_cnt, rebalance_cnt);
+
+        TEST_SAY("Awaiting leave group heartbeat\n");
+        /* After closing the consumer, 1 heartbeat should been sent */
+        TEST_ASSERT((found_heartbeats =
+                         wait_all_heartbeats_done(mcluster, 1, 200)) == 1,
+                    "Expected 1 leave group heartbeat, got %d",
+                    found_heartbeats);
+
+        rd_kafka_mock_stop_request_tracking(mcluster);
+        rd_kafka_destroy(c);
+        test_mock_cluster_destroy(mcluster);
+
+        TIMING_ASSERT(&timing, 500, 2000);
+        SUB_TEST_PASS();
+}
+
+/**
+ * @brief Test all kind of consumer fenced errors in a ConsumerGroupHeartbeat
+ *        call.
+ *        variation 0: errors on first HB
+ *        variation 1: errors on second HB
+ */
+static void do_test_consumer_group_heartbeat_fenced_errors(void) {
+        rd_kafka_resp_err_t fenced_errors[] = {
+            RD_KAFKA_RESP_ERR_UNKNOWN_MEMBER_ID,
+            RD_KAFKA_RESP_ERR_FENCED_MEMBER_EPOCH};
+        size_t i;
+        for (i = 0; i < RD_ARRAY_SIZE(fenced_errors); i++) {
+                do_test_consumer_group_heartbeat_fenced_error(fenced_errors[i],
+                                                              0);
+                do_test_consumer_group_heartbeat_fenced_error(fenced_errors[i],
+                                                              1);
+        }
+}
+
+/**
+ * @brief Test consumer group behavior with missing topic id when retrieving
+ *        metadata for assigned topics.
+ *        ensuring:
+ *        - initially a partial acknoledgement is started, with an empty list
+ *          (variation 0) or a single topic (variation 1)
+ *        - fetch doesn't start until broker returns an unknown topic id error
+ *        - when error isn't returned anymore the client finishes assigning
+ *          the partition and reads a message.
+ *
+ * @param variation See calling code.
+ */
+static void do_test_metadata_unknown_topic_id_error(int variation) {
```

Comment:
In case of UNKNOWN_TOPIC_ID_ERROR, we should eventually reach a state where we don't have that topic in the assignment list as the broker will remove that topic eventually. We should check that as well.